trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/ExtractedGhostUpdate.java

```diff
@@ -0,0 +1,214 @@
+			final ExpressionResult expressionResult) {
+		final ExpressionResult witness = instrument(loc, dispatcher);
+		final String functionName = getNameOfCalledFunction();
+		if (functionName == null) {
+			// TODO: Support other statements, also not only function calls
+			throw new UnsupportedOperationException(
+					"The following statement is not yet supported for ghost updates: " + loc);
+		}
+		switch (functionName) {
+		case "__VERIFIER_atomic_begin":
+			// Insert the ghost update after the begin of the atomic block to ensure that it is executed atomically.
+			return new ExpressionResultBuilder(expressionResult).addAllExceptLrValue(witness).build();
+		case "__VERIFIER_atomic_end":
+			// Insert the ghost update before the end of the atomic block to ensure that it is executed atomically.
+			return new ExpressionResultBuilder(witness).addAllExceptLrValue(expressionResult).build();
+		case "pthread_mutex_lock":
+			return new ExpressionResultBuilder(expressionResult).addAllExceptLrValueAndStatements(witness)
+					.resetStatements(annotateAtomicCall(loc, expressionResult.getStatements(), witness.getStatements(),
+							MemoryModelDeclarations.ULTIMATE_PTHREADS_MUTEX_LOCK.getName()))
+					.build();
+		case "pthread_mutex_unlock":
+		case "pthread_cond_wait":
+			return new ExpressionResultBuilder(expressionResult).addAllExceptLrValueAndStatements(witness)
+					.resetStatements(annotateAtomicCall(loc, expressionResult.getStatements(), witness.getStatements(),
+							MemoryModelDeclarations.ULTIMATE_PTHREADS_MUTEX_UNLOCK.getName()))
+					.build();
+		case "pthread_rwlock_rdlock":
+			return new ExpressionResultBuilder(expressionResult).addAllExceptLrValueAndStatements(witness)
+					.resetStatements(annotateAtomicCall(loc, expressionResult.getStatements(), witness.getStatements(),
+							MemoryModelDeclarations.ULTIMATE_PTHREADS_RWLOCK_READLOCK.getName()))
+					.build();
+		case "pthread_rwlock_wrlock":
+			return new ExpressionResultBuilder(expressionResult).addAllExceptLrValueAndStatements(witness)
+					.resetStatements(annotateAtomicCall(loc, expressionResult.getStatements(), witness.getStatements(),
+							MemoryModelDeclarations.ULTIMATE_PTHREADS_RWLOCK_WRITELOCK.getName()))
+					.build();
+		case "pthread_rwlock_unlock":
+			return new ExpressionResultBuilder(expressionResult).addAllExceptLrValueAndStatements(witness)
+					.resetStatements(annotateAtomicCall(loc, expressionResult.getStatements(), witness.getStatements(),
+							MemoryModelDeclarations.ULTIMATE_PTHREADS_RWLOCK_UNLOCK.getName()))
+					.build();
+		case "pthread_create":
+			// Make the ghost update itself atomic and insert it just before the fork.
+			// TODO: Maybe we should do this atomically, but the CFG builder crashes for that case
+			// We are not sure, if this does have any different semantics.
+			return new ExpressionResultBuilder(expressionResult).addAllExceptLrValueAndStatements(witness)
```

Comment:
We should look into this and see if CfgBuilder can be modified to support this; perhaps next week.

It's not a blocker for the merge though.